js/plugins/googleai/src/context-caching/utils.ts

```diff
@@ -0,0 +1,241 @@
+      message:
+        'Genkit request history and Gemini chat request history length do not match',
+    });
+  }
+}
+
+/**
+ * Looks up context cache using a cache manager and returns the found item, if any.
+ */
+export async function lookupContextCache(
+  cacheManager: GoogleAICacheManager,
+  cacheKey: string,
+  maxPages = 100,
+  pageSize = 100
+) {
+  let currentPage = 0;
+  let pageToken: string | undefined;
+
+  while (currentPage < maxPages) {
+    const { cachedContents, nextPageToken } = await cacheManager.list({
+      pageSize,
+      pageToken,
+    });
+    const found = cachedContents?.find(
+      (content) => content.displayName === cacheKey
+    );
+
+    if (found) return found;
+    if (!nextPageToken) break;
+
+    pageToken = nextPageToken;
+    currentPage++;
+  }
+  return null;
+}
+
+/**
+ * Clears all caches using the cache manager.
+ */
+export async function clearAllCaches(
+  cacheManager: GoogleAICacheManager,
+  maxPages = 100,
+  pageSize = 100
+): Promise<void> {
+  let currentPage = 0;
+  let pageToken: string | undefined;
+  let totalDeleted = 0;
+
+  while (currentPage < maxPages) {
+    try {
+      const { cachedContents, nextPageToken } = await cacheManager.list({
+        pageSize,
+        pageToken,
+      });
+      totalDeleted += await deleteCachedContents(cacheManager, cachedContents);
+
+      if (!nextPageToken) break;
+      pageToken = nextPageToken;
+      currentPage++;
+    } catch (error) {
+      throw new GenkitError({
+        status: 'INTERNAL',
+        message: `Error clearing caches on page ${currentPage + 1}: ${error}`,
+      });
+    }
+  }
+  logger.info(`Total caches deleted: ${totalDeleted}`);
+}
+
+/**
+ * Helper to delete cached contents and return the number of deletions.
+ */
+async function deleteCachedContents(
+  cacheManager: GoogleAICacheManager,
+  cachedContents: CachedContent[] = []
+): Promise<number> {
+  for (const content of cachedContents) {
+    if (content.name) await cacheManager.delete(content.name);
+  }
+  return cachedContents.length;
+}
+
+/**
+ * Extracts the cache configuration from the request if available.
+ */
+export const extractCacheConfig = (
+  request: GenerateRequest<z.ZodTypeAny>
+): {
+  cacheConfig: { ttlSeconds?: number } | boolean;
+  endOfCachedContents: number;
+} | null => {
+  const endOfCachedContents = findLastIndex<MessageData>(
+    request.messages,
+    (message) => !!message.metadata?.cache
+  );
+
+  return endOfCachedContents === -1
+    ? null
+    : {
+        endOfCachedContents,
+        cacheConfig: cacheConfigSchema.parse(
+          request.messages[endOfCachedContents].metadata?.cache
+        ),
+      };
+};
+
+/**
+ * Validates context caching request for compatibility with model and request configurations.
+ */
+export function validateContextCacheRequest(
+  request: any,
+  modelVersion: string
+): boolean {
+  if (!modelVersion || !CONTEXT_CACHE_SUPPORTED_MODELS.includes(modelVersion)) {
```

Comment:
this probably shouldn't throw - we should probably just return false

Reply:
It should probably throw if they *try* to use cache (cache is in metadata etc) but the model version is wrong. We don't want to accidentally cost them lots of tokens.